pkg/trace/api/api.go

```diff
@@ -434,50 +434,51 @@ func (r *HTTPReceiver) loop() {
 var killProcess = func(msg string) { osutil.Exitf(msg) }
 
+// watchdog checks the trace-agent's heap and CPU usage and updates the rate limiter using a correct
+// sampling rate to maintain resource usage within set thresholds. These thresholds are defined by
+// the configuration MaxMemory and MaxCPU. If these values are 0, all limits are disabled and the rate
+// limiter will accept everything.
 func (r *HTTPReceiver) watchdog(now time.Time) {
 	wi := watchdog.Info{
 		Mem: watchdog.Mem(),
 		CPU: watchdog.CPU(now),
 	}
-
-	if current, allowed := float64(wi.Mem.Alloc), r.conf.MaxMemory*1.5; current > allowed {
-		// This is a safety mechanism: if the agent is using more than 1.5x max. memory, there
-		// is likely a leak somewhere; we'll kill the process to avoid polluting host memory.
-		metrics.Count("datadog.trace_agent.receiver.oom_kill", 1, nil, 1)
-		metrics.Flush()
-		log.Criticalf("Killing process. Memory threshold exceeded: %.2fM / %.2fM", current/1024/1024, allowed/1024/1024)
-		killProcess("OOM")
-	}
-
-	rateCPU, err := sampler.CalcPreSampleRate(r.conf.MaxCPU, wi.CPU.UserAvg, r.PreSampler.RealRate())
-	if err != nil {
-		log.Warnf("Problem computing cpu pre-sample rate: %v", err)
-	}
-	rateMem, err := sampler.CalcPreSampleRate(r.conf.MaxMemory, float64(wi.Mem.Alloc), r.PreSampler.RealRate())
-	if err != nil {
-		log.Warnf("Problem computing mem pre-sample rate: %v", err)
-	}
-	r.PreSampler.SetError(err)
-	if rateCPU < 1 {
-		log.Warnf("CPU threshold exceeded (apm_config.max_cpu_percent: %.0f): %.0f", r.conf.MaxCPU*100, wi.CPU.UserAvg)
+	rateMem := 1.0
+	if r.conf.MaxMemory > 0 {
+		if current, allowed := float64(wi.Mem.Alloc), r.conf.MaxMemory*1.5; current > allowed {
+			// This is a safety mechanism: if the agent is using more than 1.5x max. memory, there
+			// is likely a leak somewhere; we'll kill the process to avoid polluting host memory.
+			metrics.Count("datadog.trace_agent.receiver.oom_kill", 1, nil, 1)
+			metrics.Flush()
+			log.Criticalf("Killing process. Memory threshold exceeded: %.2fM / %.2fM", current/1024/1024, allowed/1024/1024)
+			killProcess("OOM")
+		}
+		rateMem = computeRateLimitingRate(r.conf.MaxMemory, float64(wi.Mem.Alloc), r.RateLimiter.RealRate())
+		if rateMem < 1 {
+			log.Warnf("Memory threshold exceeded (apm_config.max_memory: %.0f bytes): %d", r.conf.MaxMemory, wi.Mem.Alloc)
+		}
 	}
-	if rateMem < 1 {
-		log.Warnf("Memory threshold exceeded (apm_config.max_memory: %.0f bytes): %d", r.conf.MaxMemory, wi.Mem.Alloc)
+	rateCPU := 1.0
+	if r.conf.MaxCPU > 0 {
+		rateCPU = computeRateLimitingRate(r.conf.MaxCPU, wi.CPU.UserAvg, r.RateLimiter.RealRate())
+		if rateCPU < 1 {
+			log.Warnf("CPU threshold exceeded (apm_config.max_cpu_percent: %.0f): %.0f", r.conf.MaxCPU*100, wi.CPU.UserAvg)
+		}
 	}
 	if rateCPU < rateMem {
-		r.PreSampler.SetRate(rateCPU)
+		r.RateLimiter.SetTargetRate(rateCPU)
 	} else {
-		r.PreSampler.SetRate(rateMem)
+		r.RateLimiter.SetTargetRate(rateMem)
 	}
```

Comment:
Nit: Since rate is already a float we could easily do `r.RateLimiter.SetTargetRate(math.Min(rateCPU, rateMem))`

Reply:
You're right. Changed it.